presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java

```diff
@@ -249,35 +254,61 @@ private Optional<PlanNode> coalesceWithNullAggregation(AggregationNode aggregati
         AggregationNode aggregationOverNull = aggregationOverNullInfo.getAggregation();
         Map<VariableReferenceExpression, VariableReferenceExpression> sourceAggregationToOverNullMapping = aggregationOverNullInfo.getVariableMapping();
 
-        // Do a cross join with the aggregation over null
-        JoinNode crossJoin = new JoinNode(
-                outerJoin.getSourceLocation(),
-                idAllocator.getNextId(),
-                JoinNode.Type.INNER,
-                outerJoin,
-                aggregationOverNull,
-                ImmutableList.of(),
-                ImmutableList.<VariableReferenceExpression>builder()
-                        .addAll(outerJoin.getOutputVariables())
-                        .addAll(aggregationOverNull.getOutputVariables())
-                        .build(),
-                Optional.empty(),
-                Optional.empty(),
-                Optional.empty(),
-                Optional.empty(),
-                ImmutableMap.of());
+        FunctionResolution functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+        Map<VariableReferenceExpression, RowExpression> literalMap = new HashMap<>();
+
+        if (useDefaultsForCorrelatedAggregations) {
+            for (Map.Entry<VariableReferenceExpression, AggregationNode.Aggregation> aggregation : aggregationNode.getAggregations().entrySet()) {
+                FunctionHandle functionHandle = aggregation.getValue().getFunctionHandle();
+
+                RowExpression defaultLiteral = null;
+                if (functionResolution.isCountFunction(functionHandle)) {
```

Comment:
I think count(column) over a single null row returns 0, but count(*) returns 1.  did a simple test to try it out.
```
 SELECT count(x) from ( SELECT null as x from dim_one_row);
 _col0                
-------
     0 
(1 row)

SELECT count(*) from ( SELECT null as x from dim_one_row);
 _col0                
-------
     1 
(1 row)
```

Reply:
While this is true, I'm not sure it is relevant. At least I haven't been able to come up with a case that breaks the implementation. 
Previously we were adding the cross join in order to coalesce the nulls produced by the outer join to appropriate aggregations over nulls. With this change, the outer join gets converted into an inner anyway (due to the introduction of a null-intolerant predicate) and so cannot produce nulls. The only remaining question is whether the correlated count in the subquery produces a single null row as output - but given that it is a count aggregation (with * or column as input), this can never happen. So I do not think the behavior of count(null) is relevant in this case. Thoughts?